ring/consul_client.go

```diff
@@ -64,25 +72,112 @@ var (
 	ErrNotFound = fmt.Errorf("Not found")
 )
 
-// Get and deserialise a JSON value from Consul.
-func (c *consulClient) Get(key string, factory InstanceFactory) error {
-	kvp, _, err := c.kv.Get(key, queryOptions)
+// ProtoSerDes is a SerDes for proto/snappy
+type ProtoSerDes struct {
+	Factory func() proto.Message
+}
+
+// Decode implements SerDes
+func (p ProtoSerDes) Decode(bytes []byte) (interface{}, error) {
+	out := p.Factory()
+	bytes, err := snappy.Decode(nil, bytes)
+	if err != nil {
+		return nil, err
+	}
+	if err := proto.Unmarshal(bytes, out); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// Encode implements SerDes
+func (p ProtoSerDes) Encode(msg interface{}) ([]byte, error) {
+	bytes, err := proto.Marshal(msg.(proto.Message))
 	if err != nil {
-		return err
+		return nil, err
+	}
+	return snappy.Encode(nil, bytes), nil
+}
+
+// JSONSerDes is a SerDes for JSON
+type JSONSerDes struct {
+	Factory func() interface{}
+}
+
+// Decode implements SerDes
+func (j JSONSerDes) Decode(bytes []byte) (interface{}, error) {
+	out := j.Factory()
+	if err := json.Unmarshal(bytes, out); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// Encode implemenrs SerDes
+func (j JSONSerDes) Encode(msg interface{}) ([]byte, error) {
+	return json.Marshal(msg)
+}
+
+// DynamicSerDes is a SerDes that can read json and proto, and
+// that can serialise to either (selectively).
+// Once it fails to decode JSON, it will start decoding (and
+// writing) protos.
+type DynamicSerDes struct {
+	mtx      sync.Mutex
+	useProto bool
+	json     SerDes
+	proto    SerDes
+}
+
+// NewDynamicSerDes makes a new DynamicSerDes
+func NewDynamicSerDes(json, proto SerDes) *DynamicSerDes {
+	return &DynamicSerDes{
+		useProto: false,
+		json:     json,
+		proto:    proto,
```

Comment:
A different approach would be to have a vector of `SerDes` instead of special `json` and `proto` ones, and an index of the currently selected format, instead of `useProto`. That way this would work with future format changes.

Then the public `UseProto` would become `UseVersion` and the check for `allIngestersCanReadProtos` would instead find the minimum supported version.

Not sure it's worth it though. Would help for future compatibility stuff and for tests (because you could pass much simpler codecs & objects).